src/Features/CSharp/Portable/UseCompoundAssignment/CSharpUseCompoundAssignmentCodeFixProvider.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.UseCompoundAssignment;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCompoundAssignment
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    internal class CSharpUseCompoundAssignmentCodeFixProvider 
+        : AbstractUseCompoundAssignmentCodeFixProvider<SyntaxKind, AssignmentExpressionSyntax, ExpressionSyntax>
+    {
+        public CSharpUseCompoundAssignmentCodeFixProvider()
+            : base(Utilities.Kinds)
+        {
+        }
+
+        protected override SyntaxKind GetSyntaxKind(int rawKind)
```

Comment:
all these seems can be moved to ISyntaxFactService? that contains a lot of these kind of methods already.

Reply:
I do not see how.  If you look at the conversation, i attempted to do that already, but found that i could not.  If you have a way to actually move it, let me know!

Reply:
could you also clarify what you mean by "all" in this context?  do you jus tmean this method, or the other methods that follow as well?

Reply:
I was thinking something like
ISyntaxFactService.GetSyntaxKind(int rawKind) and ISyntaxFactService.Token(int rawKind) and AssignmentExpression(int kind, node left, token token, node right)

actually, last one might candidate to add to SyntaxGenerator?

Reply:
> ISyntaxFactService.GetSyntaxKind(int rawKind) 

But what would be the return type of GetSyntaxKind be?  it can't be 'int', so that wuold be unhelpful.  If we made GetSyntaxKind generic on TSyntaxKind that would be super strange to have to implement in a cheap fashion :)

>  `ISyntaxFactService.Token(int rawKind)`.  

But it's not a SyntaxFact :)  it's generating a Token :)

> actually, last one might candidate to add to SyntaxGenerator?

I did think about htat, however, it doesn't really fix SyntaxGenerator (which is about a language neutral way of generating stuf that does not deal with things like kinds/tokens). If we did that, i'd expect we'd need "token" accepting forms for most of the code in there...

Reply:
the fact that it is currently return SyntaxKind means the base know the type even though it might be generic so.
TSyntaxKind GetSyntaxKind<TSyntaxKind>(int rawKind) something like that?

> But it's not a SyntaxFact :) it's generating a Token :)
I see in ISyntaxFactService it sometimes return new token or modify node or etc. but if that location is problem, it could be in SyntaxGenerator.

anyway, end result I was looking for was removing these language specific types here since it seems once these 3 can move to those service, these are not needed.

Reply:
> `TSyntaxKind GetSyntaxKind(int rawKind)` something like that?

That feels very weird to me.  Since hte impl will effectively have to do something like `return (TSyntaxKind)(object)(rawKind)`.  I'd prefer to avoid the potential for boxing here, unless it can be strongly guaranteed that the runtime will do the right thing here.

>  but if that location is problem, it could be in SyntaxGenerator.

I could see `Token(rawKind)` being in generator.  I don't think AssignmentExpression should go there.

Reply:
> anyway, end result I was looking for was removing these language specific types here since it seems once these 3 can move to those service, these are not needed.

Sure, and in general i'm supportive of that. As you can see in the PR, several additions to ISyntaxFacts were made.  However, the general purposes of these services should be followed.  They're not intended to be a grab bag where anything language specific can go (and i've tried ot prevent people from using them as such over the years, though not alwyas successfully).  So, for example, ISyntaxFactsService is supposed to just be for queries.  And we should try to keep it htat way (removing inappropriate stuff if necessary).  SyntaxGenerator should be for specifically providing an language-neutral abstraction over general cross language concepts.  That's why it never deals with kinds, and only ever really deals with tokens for things like an identifier.  I'd like to preserve those general principles.

In other words, even if we could do this, that doesn't mean we should :)